tools/secret-rotation/Azure.Sdk.Tools.SecretRotation.Stores.AzureDevOps/ServiceAccountPersonalAccessTokenStore.cs

```diff
@@ -0,0 +1,262 @@
+        this.patDisplayName = patDisplayName;
+        this.scopes = scopes;
+        this.serviceAccountTenantId = serviceAccountTenantId;
+        this.serviceAccountName = serviceAccountName;
+        this.serviceAccountPasswordSecret = serviceAccountPasswordSecret;
+        this.revocationAction = revocationAction;
+        this.tokenCredential = tokenCredential;
+        this.secretProvider = secretProvider;
+    }
+
+    public override bool CanOriginate => true;
+
+    public override bool CanRevoke => true;
+
+    public static Func<StoreConfiguration, SecretStore> GetSecretStoreFactory(TokenCredential credential, ISecretProvider secretProvider, ILogger logger)
+    {
+        return configuration =>
+        {
+            var parameters = configuration.Parameters?.Deserialize<Parameters>();
+
+            if (string.IsNullOrEmpty(parameters?.Organization))
+            {
+                throw new Exception("Missing required parameter 'organization'");
+            }
+
+            if (string.IsNullOrEmpty(parameters.PatDisplayName))
+            {
+                throw new Exception("Missing required parameter 'patDisplayName'");
+            }
+
+            if (string.IsNullOrEmpty(parameters.Scopes))
+            {
+                throw new Exception("Missing required parameter 'scopes'");
+            }
+
+            if (string.IsNullOrEmpty(parameters.ServiceAccountTenantId))
+            {
+                throw new Exception("Missing required parameter 'serviceAccountTenantId'");
+            }
+
+            if (string.IsNullOrEmpty(parameters.ServiceAccountName))
+            {
+                throw new Exception("Missing required parameter 'serviceAccountName'");
+            }
+
+            if (string.IsNullOrEmpty(parameters.ServiceAccountPasswordSecret))
+            {
+                throw new Exception("Missing required parameter 'serviceAccountPasswordSecret'");
+            }
+
+            if (!Uri.TryCreate(parameters.ServiceAccountPasswordSecret, UriKind.Absolute, out Uri? serviceAccountPasswordSecret))
+            {
+                throw new Exception("Unable to parse Uri from parameter 'serviceAccountPasswordSecret'");
+            }
+
+            return new ServiceAccountPersonalAccessTokenStore(
+                logger,
+                parameters.Organization,
+                parameters.PatDisplayName,
+                parameters.Scopes,
+                parameters.ServiceAccountTenantId,
+                parameters.ServiceAccountName,
+                serviceAccountPasswordSecret!,
+                parameters.RevocationAction,
+                credential,
+                secretProvider);
+        };
+    }
+
+    public override async Task<SecretValue> OriginateValueAsync(SecretState currentState, DateTimeOffset expirationDate, bool whatIf)
+    {
+        VssConnection connection = await GetConnectionAsync();
+        TokensHttpClient client = connection.GetClient<TokensHttpClient>();
```

Comment:
Same comment above re: whatIf scope.

Reply:
same here.  whatif should do everything but the write.